Expose final Helm values used to render the chart in a ConfigMap

luksa opened the pull request:
**Please provide a description of this PR:**

Before profiles were implemented within the helm charts, users were able to see the entire set of values that were used to render the chart by running `helm get values`. This is no longer possible, since the user profile, platform profile, and compatibility version values are applied during rendering. There is no way to get the final set of values. Some tools (e.g. the Sail operator) need to know the final set of values, as do some users.

This change adds a Configmap called `values` or `values-<revision>` that tools/users can inspect to see what values were actually used to render the templates.

luksa committed:
Expose final Helm values used to render the chart

Before profiles were implemented within the helm charts, users were able to see the entire set of values that were used to render the chart by running `helm get values`. This is no longer possible, since the user profile, platform profile, and compatibility version values are applied during rendering. There is no way to get the final set of values. Some tools (e.g. the Sail operator) need to know the final set of values, as do some users.

This change adds a Configmap called `values` or `values-<revision>` that tools/users can inspect to see what values were actually used to render the templates.

istio-testing commented:
@luksa: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | fc3346a20103a490140f80fb24acac3d7c583536 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/release-notes_istio/1908120957558460416) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

luksa commented:
I'll add tests after folks confirm that this change is reasonable.  If it is, does there have to be a way to disable the creation of this ConfigMap?

istio-testing commented:
@luksa: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | fc3346a20103a490140f80fb24acac3d7c583536 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/release-notes_istio/1908120957558460416) | true | `/test release-notes`
integ-pilot-multicluster_istio | fc3346a20103a490140f80fb24acac3d7c583536 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/integ-pilot-multicluster_istio/1908120948129665024) | true | `/test integ-pilot-multicluster`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn reviewed (commented):
One worry I have is users see this and expect that they can modify it to make changes. We saw this many many times when we used to do this with IstioOperator.

maybe we can encode it (base64 or such) to avoid this?

luksa committed:
Rename field to last-applied-values

luksa committed:
Add description

istio-testing commented:
@luksa: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-multicluster_istio | fc3346a20103a490140f80fb24acac3d7c583536 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/integ-pilot-multicluster_istio/1908120948129665024) | true | `/test integ-pilot-multicluster`
release-notes_istio | ad571c47623134162c48c6a530d76c57aa9a3f32 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/release-notes_istio/1909137705069973504) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

luksa commented:
@howardjohn You're right. We need to make it clear that the values shouldn't be changed.

Not sure about base64 encoding the values, since that would just make the UX slightly worse, and users may still end up thinking that they can modify it.

I've added the `kubernetes.io/description` annotation to make the point of the ConfigMap perfectly clear. I've also renamed the item key from `values` to `last-applied-values`. Do you think the ConfigMap itself should also be renamed?

Side-note: do you think it would make sense to add description annotations to other resources as well? At least ConfigMaps and Secrets. It would be kinda nice for users to see what each resource's role is without having to check the docs.

istio-testing commented:
@luksa: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | ad571c47623134162c48c6a530d76c57aa9a3f32 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/release-notes_istio/1909137705069973504) | true | `/test release-notes`
integ-cni_istio | ad571c47623134162c48c6a530d76c57aa9a3f32 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/integ-cni_istio/1909137694454190080) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
> @howardjohn You're right. We need to make it clear that the values shouldn't be changed.
> 
> Not sure about base64 encoding the values, since that would just make the UX slightly worse, and users may still end up thinking that they can modify it.
> 
> I've added the `kubernetes.io/description` annotation to make the point of the ConfigMap perfectly clear. I've also renamed the item key from `values` to `last-applied-values`. Do you think the ConfigMap itself should also be renamed?
> 
> Side-note: do you think it would make sense to add description annotations to other resources as well? At least ConfigMaps and Secrets. It would be kinda nice for users to see what each resource's role is without having to check the docs.

Hmm, I have never heard of the annotation but it does seem like it could be useful. I don't think it will do to much to stop users from modifying it, the users that are inclined to modify random configmaps don't read warnings :stuck_out_tongue: 

Re: UX, I was thinking this was mostly tools consuming so not a big deal, but re-reading the PR I think that is not the only use case.

One thing we see users say is "<previous coworker> installed Istio and I have no clue how they did it". This almost solves this use case, but I wonder if it will lead them down the wrong path by giving them the fully merged path. Then they will try to re-install with `-f full-values.yaml` when they really want `-f original-values.yaml`. Similar-ish for istioctl where they really want `-f istio-operator.yaml`.

Does it make sense to have 2 keys, original and merged or something? I don't mind not doing it, just a thought

luksa committed:
Expose both original and merged values

istio-testing commented:
@luksa: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-cni_istio | ad571c47623134162c48c6a530d76c57aa9a3f32 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/integ-cni_istio/1909137694454190080) | true | `/test integ-cni`
release-notes_istio | 228439f0d0478c236df8fbc3934d36b20773d384 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/55803/release-notes_istio/1909640746479128576) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

luksa commented:
@howardjohn I've now added the original values as well. But we should probably create a separate ConfigMap for every chart. Not ideal.

A few thoughts:
- Should we include `-readonly` in the ConfigMap name?
- Is `values` clear enough or should it be called `helm-values` or something similar?
- Should we create the ConfigMap in `zzz_profile.yaml` directly? This way we wouldn't have to pass the original values through `.Values._original`.

dgn reviewed (approved):
LGTM. Probably worth adding a release note

dgn commented on manifests/charts/istio-control/istio-discovery/templates/configmap-values.yaml:
```diff
@@ -0,0 +1,21 @@
+  name: values{{- if not (eq .Values.revision "") }}-{{ .Values.revision }}{{- end }}
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    kubernetes.io/description: This ConfigMap contains the Helm values used during chart rendering. Modifying
```
maybe 
```suggestion
    kubernetes.io/description: This ConfigMap contains the Helm values used during chart rendering. This ConfigMap is rendered for debugging purposes and external tooling: modifying these values has no effect.
```

luksa committed:
Improve description

luksa committed:
Add release note
